src/main/java/teammates/common/util/AppUrl.java

```diff
@@ -0,0 +1,49 @@
+package teammates.common.util;
+
+/**
+ * A specific implementation of {@link Url} used to encapsulate URLs of the application.
+ */
+public class AppUrl extends Url {
+    
+    private static final String BASE_WEB_URL_REGEX = "^(https?://[^/]+)";
+    
+    public AppUrl(String baseUrl, String url) {
+        super(baseUrl, makeRelativeUrl(url)); // force the stored URL to be relative
+        Assumption.assertTrue(baseUrl.matches(BASE_WEB_URL_REGEX + "$"));
+        String relativeUrl = makeRelativeUrl(url);
+        Assumption.assertTrue(relativeUrl.isEmpty() || relativeUrl.startsWith("/"));
```

Comment:
1. This processing is not specific to AppUrl. Shouldn't it be pushed to parent class?
2. Why does this constructor need a baseUrl parameter? Given a full URL, can't we automatically extract the base from it? If we can, I'd rather have two constructors `(fullUrl)` and `(baseUrl, relativePath)`.
3. Add header comments. Specify if the parameters require a trailing slash (e.g. it works with or without the trailing slash but the outputs will not have trailing slashes, in which case the trimming should happen here)